Windows: Neovim-Qt: missing QtSVG DLL

jgehrig opened the pull request:
**Issue #12928:** Windows titlebar icon not displayed 

We should copy the entire output of `windeployqt`. The issue is caused by a missing `qsvgd.dll`, but there are several other missing dependencies.

Currently we whitelist a subset of files. Instead, we should perform a recursive copy for all `windeployqt` dependencies.

teto commented:
we need windows user to approve. cc @erw7 ?

jgehrig commented:
Not ready yet :)

The general location of the fix is correct, but the syntax is currently invalid. I need to set up a Windows machine.

Sorry for the early post. I wanted to validate against your AppVeyor CI.

teto commented:
ok my bad, I've prefix the PR [WIP], change it to [RDY] (ready) when done.

Shatur95 commented:
This is a bit hacky way :)

I suggest you to use built-in [windeployqt.exe](https://doc.qt.io/qt-5/windows-deployment.html):

```cmake
    find_program(WINDEPLOYQT windeployqt)
    install(CODE "execute_process(COMMAND ${WINDEPLOYQT} \${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_BINDIR}/$<TARGET_FILE_NAME:${PROJECT_NAME}>)")
```

I use it in my [Crow Translate](https://github.com/crow-translate/crow-translate/blob/9adb63f03b1ddbf4749149a3f16c0156667d3dcf/CMakeLists.txt#L181) (it also requires QtSVG and these 2 lines works for me).

jgehrig commented:
@Shatur95 

I completely forgot about this!

We already use `windeployqt`, but the cmake logic in neovim selectively copies files. We need to copy all of the output, not just a whitelist of files.

Yeah, my CMake knowledge is a bit lacking...

Shatur95 commented:
Yes, this would be safer than just hard-coding the list of required files.

jgehrig commented:
Well, I'm lost... The fix should be simple: copy the full output of `windeployqt` via `GLOB_RECURSE`.

However, I can't figure out how to make Windows build Neovim. Using AppVeyor was also unsuccessful, pull requests only trigger a partial build and do not include `third-party` or `neovim-qt`.

If anyone can provide assistance it would be greatly appreciated. Otherwise, I'm out of ideas.

jgehrig committed:
Issue 12928: Windows missing QtSVG DLL

We should copy the entire output of windeployqt. Currently several DLLs
are missing including one required to display the window icon.

jgehrig commented:
@teto

I've taken the easy approach, we will continue to hard-code the necessary DLLs.

My original attempt to use a recursive copy doesn't work because `neovim-qt` build output goes to the same directory as other build dependencies.

I built the changes locally on Windows, everything looks good. A clean copy of `master` still exhibits the icon issues. When these changes are applied, the window icon is rendered properly.

jgehrig commented:
Ping @teto @erw7

This change is ready.

It'd be great to get this checked-in. I've seen several different bug reports for this issue. Thanks!

jgehrig commented:
@jamessan

Do you know who's approval is needed for merge?

This is a simple fix to resolve the icon issues for the packaged version of neovim-qt on Windows. Some required DLLs are missing. This is tested, and to the best of my knowledge ready for check-in.

Thanks!

erw7 commented:
I'm sorry I didn't have time. I confirmed that it looks fine.

Just wondering, do we need to bundle `neovim-qt` into the Windows package? Previously, `tui` didn't work on Windows, I think that's why it shipped with `neovim-qt`. However, `tui` now works on Windows as well. Therefore, I don't think there is any point in bundling `neovim-qt` anymore.

janlazo commented:
@erw7 We don't have to but neovim-qt links to the neovim wiki for its install instructions. neovim-qt wiki should be updated to include its own install instructions in its README.md or https://github.com/equalsraf/neovim-qt/wiki#faq so that neovim can link to it.

jgehrig commented:
Sorry for the noise everyone. Thanks @erw7!

Personally, I'd like to see the bundling of neovim-qt continue. I think the terminal experience on Windows (in general) is poor. I suspect bundling is useful for many Windows users. Mouse input is especially awkward in the terminal on Windows.

There isn't currently a good set of install instructions on Windows... The current process on Windows is manual, so I would need to come up with something new.

A README update is pending for neovim-qt: https://github.com/equalsraf/neovim-qt/pull/821

I can also update the Wiki with any content you'd like.

erw7 commented:
@jgehrig merged, thanks!

erw7 commented:
> Personally, I'd like to see the bundling of neovim-qt continue. I think the terminal experience on Windows (in general) is poor. I suspect bundling is useful for many Windows users. Mouse input is especially awkward in the terminal on Windows.

I think terminal support in Windows has improved a lot compared to the past. The mouse should work even if it is cmd.exe (conhost.exe) as well as the Windows terminal.

Currently, the neovim-qt release is only available in 32-bit version. I think the 64bit dll of `msvcrt.dll` or something was overwritten by the 32bit dll of neovim-qt, so we should stop bundling it. Now I see that msvcrt.dll is not included in the package, so I may have been mistaken.

teto commented:
removing neovim-qt right before 0.5 may not be the best call but we can prepare for a future removal by moving the instructions to neovim-qt wiki as previously mentioned.

jgehrig commented:
I'd still like to see a GUI bundled with Neovim on Windows, but that's just my two cents...

Agreed. The terminal experience on Windows has improved in recent years. However, I still find it lacking when compared to Mac/Linux. Unfortunately, plenty of users are still running "cmd.exe" and not the new "Windows Terminal". I think there is value in providing a "just works" GUI solution; especially on Windows where new users are more likely to struggle with proper `%PATH%` configuration between nvim and frontend.

Please @-mention me when/if you drop the bundled version. I will update the Wiki/README accordingly with a final bundled version and alternative install instructions. I'll keep an eye on this and make sure the documentation doesn't go stale.

jdhao commented:
I am using Windows on my company PC. Although I am using Windows 10, the version is not up-to-date. So I cannot use windows terminal. Other terminals on Windows I have tried so far does not work well. The nvim tui has various issues on these terminals (font issue, color issue, glitches etc.)

The best nvim experience we can get is from nvim-qt. I think it should not be removed.

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -513,23 +513,58 @@ if(WIN32)
+                      translations/qt_pl.qm
+                      translations/qt_ru.qm
+                      translations/qt_sk.qm
+                      translations/qt_uk.qm
```
are these actually used/needed?

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -513,23 +513,58 @@ if(WIN32)
+                      imageformats/qwbmp.dll
+                      imageformats/qwebp.dll
+                      platforms/qwindows.dll
+                      styles/qwindowsvistastyle.dll
```
is this actually needed?

justinmk commented:
> Currently we whitelist a subset of files. Instead, we should perform a recursive copy for all `windeployqt` dependencies.

Reason for excluding some DLLs is that some of them are big and it bloats the `nvim-win64.zip` download.

jgehrig commented:
The files here are the full set reported by `windeployqt`. If a DLL is included, there should be usage in the `neovim-qt` code somewhere.

With that said, I can remove any of these DLLs if desired. Clearly we were ran `nvim-qt` fine before the addition of these files. The only file causing a known issue was `qsvgd.dll` and the missing icon.

Here is my understanding of each file/dll you mentioned:
`imageformats/*.dll` - Image rendering formats, we only use SVG icons. Not sure what the dependency graph looks like.
`platforms/qwindows.dll` - Qt Windows platform plugin. I think we should keep this.
`styles/*.dll` - Optional Qt theming/styles. Could be removed if desired.
`translations/*` - Translation files for calls to `tr(...)`. For example, `--help` can display in other languages.
